packages/remix-react/components.tsx

```diff
@@ -1296,7 +1296,45 @@ export function useMatches() {
  *
  * @see https://remix.run/api/remix#useloaderdata
  */
-export function useLoaderData<T = AppData>(): T {
+
+// matches any function
+export type TypedResponse<T> = Response & {
+  json(): Promise<T>;
+};
+type DataFunction = (...args: any[]) => any;
+type DataOrFunction = AppData | DataFunction;
+type JsonPrimitives = string | number | boolean | null;
+type NonJsonPrimitives = undefined | Function | symbol;
+type SerializeType<T> = T extends JsonPrimitives
+  ? T
+  : T extends undefined
+  ? undefined
+  : T extends { toJSON(): infer U }
+  ? U
+  : T extends []
+  ? []
+  : T extends [any, ...any[]]
+  ? {
+      [k in keyof T]: T[k] extends NonJsonPrimitives
+        ? null
+        : SerializeType<T[k]>;
+    }
+  : T extends (infer U)[]
+  ? SerializeType<U>[]
+  : {
+      [k in keyof T as T[k] extends NonJsonPrimitives
+        ? never
+        : k]: SerializeType<T[k]>;
+    };
+
+export type UseDataFunction<T extends DataOrFunction> = T extends (
```

Comment:
Shouldn't this be called `UseDataFunctionReturn`, since it's the return type of `useActionData` & `useLoaderData`?

Reply:
Changed to your recommendation